File: apps/processor-service/src/processor.controller.ts
import {
  Controller,
  Post,
  Get,
  Body,
  Param,
  ParseIntPipe,
  HttpCode,
  HttpStatus,
} from '@nestjs/common';
import {
  ApiTags,
  ApiOperation,
  ApiResponse,
  ApiParam,
} from '@nestjs/swagger';
import { ProcessorService } from './processor.service';
import { TransferDto } from './dto/transfer.dto';

@ApiTags('Processor')
@Controller()
export class ProcessorController {
  constructor(private readonly processorService: ProcessorService) {}

  @Get('health')
  @ApiOperation({ summary: 'Health check del servicio' })
  @ApiResponse({ status: 200, description: 'Servicio activo' })
  health() {
    return {
      status: 'ok',
      service: 'processor-service',
      timestamp: new Date().toISOString(),
    };
  }

  @Post('api/transfer')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({
    summary: 'Transferir dinero entre usuarios (Patrón Saga)',
    description:
      'Orquesta una transferencia P2P usando el Patrón Saga. ' +
      'Flujo: PENDING → (debit sender) → DEBITED → (credit receiver) → COMPLETED. ' +
      'Si el crédito falla, compensa con credit al sender → ROLLED_BACK.',
  })
  @ApiResponse({
    status: 200,
    description: 'Transferencia completada exitosamente (status: COMPLETED)',
  })
  @ApiResponse({
    status: 400,
    description:
      'Datos inválidos: auto-transferencia (self_transfer_not_allowed) o monto <= 0 (invalid_amount)',
  })
  @ApiResponse({
    status: 409,
    description: 'Fondos insuficientes en accounts-service (insufficient_funds)',
  })
  @ApiResponse({
    status: 422,
    description:
      'Crédito falló pero compensación exitosa: dinero devuelto al sender (ROLLED_BACK)',
  })
  @ApiResponse({
    status: 503,
    description: 'Accounts Service no disponible',
  })
  async transfer(@Body() dto: TransferDto) {
    return this.processorService.transfer(dto);
  }

  @Get('api/transactions/:user_id')
  @ApiOperation({
    summary: 'Historial de transacciones de un usuario',
    description:
      'Retorna todas las transacciones donde el usuario fue sender o receiver, ordenadas por fecha descendente.',
  })
  @ApiParam({ name: 'user_id', type: 'number', description: 'ID del usuario' })
  @ApiResponse({
    status: 200,
    description: 'Lista de transacciones con campo "type": sent | received',
  })
  async getTransactionHistory(
    @Param('user_id', ParseIntPipe) userId: number,
  ) {
    return this.processorService.getTransactionHistory(userId);
  }
}

File: neowallet/processor-service/src/main/java/com/neowallet/processor/infrastructure/adapter/in/web/ProcessorController.java
package com.neowallet.processor.infrastructure.adapter.in.web;

import com.neowallet.processor.domain.port.in.GetTransactionsUseCase;
import com.neowallet.processor.domain.port.in.TransferUseCase;
import com.neowallet.processor.infrastructure.adapter.in.web.dto.TransactionResponse;
import com.neowallet.processor.infrastructure.adapter.in.web.dto.TransferRequest;
import com.neowallet.processor.infrastructure.adapter.in.web.dto.TransferResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api")
@Tag(name = "Processor", description = "Procesamiento de transferencias P2P")
public class ProcessorController {

    private final TransferUseCase transferUseCase;
    private final GetTransactionsUseCase getTransactionsUseCase;

    public ProcessorController(TransferUseCase transferUseCase,
                                GetTransactionsUseCase getTransactionsUseCase) {
        this.transferUseCase = transferUseCase;
        this.getTransactionsUseCase = getTransactionsUseCase;
    }

    // ─────────────────────────────────────────────
    // RF-003: POST /api/transfer
    // ─────────────────────────────────────────────
    @PostMapping("/transfer")
    @Operation(
        summary = "Transferir dinero P2P (RF-003)",
        description = "Ejecuta una transferencia entre dos usuarios usando el patrón Saga con compensación",
        security = @SecurityRequirement(name = "bearerAuth")
    )
    @ApiResponses({
        @ApiResponse(responseCode = "200", description = "Transferencia exitosa"),
        @ApiResponse(responseCode = "400", description = "Fondos insuficientes, auto-transferencia, o monto inválido"),
        @ApiResponse(responseCode = "404", description = "Usuario no encontrado"),
        @ApiResponse(responseCode = "503", description = "Error de comunicación con Accounts Service")
    })
    public ResponseEntity<TransferResponse> transfer(@Valid @RequestBody TransferRequest request) {
        var transaction = transferUseCase.transfer(
            request.senderId(),
            request.receiverId(),
            request.amount()
        );
        return ResponseEntity.ok(TransferResponse.from(transaction));
    }

    // ─────────────────────────────────────────────
    // RF-005: GET /api/transactions/{userId}
    // ─────────────────────────────────────────────
    @GetMapping("/transactions/{userId}")
    @Operation(
        summary = "Historial de transacciones (RF-005)",
        description = "Retorna todas las transacciones donde el usuario fue sender o receiver",
        security = @SecurityRequirement(name = "bearerAuth")
    )
    @ApiResponses({
        @ApiResponse(responseCode = "200", description = "Lista de transacciones"),
        @ApiResponse(responseCode = "401", description = "No autenticado")
    })
    public ResponseEntity<List<TransactionResponse>> getTransactions(
            @Parameter(description = "ID del usuario", example = "1")
            @PathVariable Long userId) {
        var transactions = getTransactionsUseCase.getTransactions(userId);
        var response = transactions.stream()
                .map(t -> TransactionResponse.from(t, userId))
                .toList();
        return ResponseEntity.ok(response);
    }
}
